### PR TITLE
optimize processing of endpoints when reacting to rule updates

### PR DESCRIPTION
**Pull request by ianvernon**
The sets of endpoints which need to be regenerated by a given policy update vs.
those which need to have their revision bumped directly are always disjoint.
Furthermore, when evaluating whether a set of rules select a given endpoint, the
set of endpoints which needs to be regenerated can be updated, and no more rules
need to be evaluated, as the set of endpoints to be regenerated has already been
updated for said endpoint. This means that said endpoint can be removed from the
original set containing all endpoints. The current flow when reacting to rule
updates is as follows:

* Add all endpoints to the set of endpoints which need to have their revision
bumped
* If an endpoint needs to be regenerated because rules which were added,
deleted, or updated which select the endpoint, then remove the endpoint from the
set of all endpoints and add it to the set of endpoints which need to be
regenerated. No further processing needs to be performed for this endpoint apart
from regenerating it later.
* After all endpoints are processed in relation to all rule changes, the set of
endpoints which formerly contained all endpoints will eventually contain only
the endpoints whose revision needs to be bumped.

This makes the reacting to rule updates more performant, as the iteration of
rules for a given endpoint is no longer always `O(n)`, but at best case may be
`O(1)`.

Furthermore, an endpoint is never evaluated again for rule matching (which is
costly) as part of the reacting after it is determined to be needed to be
regenerated.

A side-effect of this is that the updating of whether rules select the endpoint
is deferred to occur when the endpoint is being regenerated, but this will only
happen once per-rule per-endpoint. This is OK because the caches within the
rules are lazily updated.

Signed-off by: Ian Vernon <ian@cilium.io>

Related-to: #7466 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7465)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

File: pkg/policy/identifier.go [hunk 1]
```diff
@@ -42,3 +42,19 @@ type Endpoint interface {
 	GetSecurityIdentity() *identity.Identity
 	PolicyRevisionBumpEvent(rev uint64)
 }
+
+// EndpointSet is used to be able to group together a given set of Endpoints
+// that need to have a specific operation performed upon them (e.g., policy
+// revision updates).
+type EndpointSet struct {
+	Mutex     lock.RWMutex
```
Comment (tgraf) [inline pkg/policy/identifier.go]: I would prefer for the Mutex to not be exposed but instead having `Add()`, `Delete()`, and `ForEach()` functions which hides the locking.

**Review by tgraf — CHANGES_REQUESTED** [review]
The change itself looks good. I would prefer if the Mutex was not public but I'm fine to merge as-is as well if this is not simple to change given the upcoming code freeze.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/policy/identifier.go (hunk 1 above)
Comment (ianvernon) [reply pkg/policy/identifier.go]: @tgraf done

**Comment by ianvernon** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]
Looks very clean now!

**Pull request merged**